### ISSUE TITLE
Zombie reaping logic hijacks child exit status

### ISSUE THREAD

**Issue by zcahana** [amalgam8/amalgam8#401]
The newly added zombie reaping logic reacts to SIGCHLD by doing a wait() syscall.
This hijacks the child's exit status from competing wait() calls, interfering with......

1. Nginx updater code which [fails to verify that nginx is running](https://github.com/amalgam8/amalgam8/blob/master/sidecar/proxy/nginx/service.go#L82), so doesn't update nginx with new rules/services:
```
time="2016-11-17T12:52:44Z" level=error msg="Could not get status of NGINX service" error="waitid: no child processes" 
time="2016-11-17T12:52:44Z" level=warning msg="Controller listener failed" error="waitid: no child processes" 
```

2. Command-type healthchecks (e.g., `file:///bin/ls` for the productpage microservice) which [fails to read the command exit status](https://github.com/amalgam8/amalgam8/blob/master/sidecar/register/healthcheck/command.go#L87
), and triggers deregistration:
```
time="2016-11-17T12:46:04Z" level=info msg="Starting Amalgam8 service registration agent" service_name=productpage 
time="2016-11-17T12:46:04Z" level=info msg="Service successfully registered with Amalgam8" instance_id=c366a7b3223239bc service_name=productpage 
time="2016-11-17T12:48:29Z" level=info msg="Stopping Amalgam8 service registration agent" service_name=productpage 
time="2016-11-17T12:48:29Z" level=info msg="Attempting to deregister service with Amalgam8" instance_id=c366a7b3223239bc service_name=productpage 
time="2016-11-17T12:48:29Z" level=info msg="Service successfully deregistered with Amalgam8" instance_id=c366a7b3223239bc service_name=productpage 
time="2016-11-17T12:48:59Z" level=info msg="Starting Amalgam8 service registration agent" service_name=productpage 
time="2016-11-17T12:48:59Z" level=info msg="Service successfully registered with Amalgam8" instance_id=c366a7b3223239bc service_name=productpage 
time="2016-11-17T12:49:29Z" level=info msg="Stopping Amalgam8 service registration agent" service_name=productpage 
time="2016-11-17T12:49:29Z" level=info msg="Attempting to deregister service with Amalgam8" instance_id=c366a7b3223239bc service_name=productpage 
time="2016-11-17T12:49:29Z" level=info msg="Service successfully deregistered with Amalgam8" instance_id=c366a7b3223239bc service_name=productpage

**Comment by rshriram** [comment]
On a slightly related note : Is the new logic capable of handling this issue #62 ?

**Comment by zcahana** [comment]
The zombie reaping logic itself is only concerned with, well, zombie reaping - it doesn't start/restart any processes.

The supervise package indeed handles process supervision... but per @GregHanson's [comment](https://github.com/amalgam8/amalgam8/issues/62#issuecomment-258144447), splitting the nginx management logic between two components is not desired.

**Comment by GregHanson** [comment]
@zcahana can you document the exact steps to reproduce the nginx item?

**Comment by zcahana** [comment]
This is intermittent (race condition), so doesn't reproduce consistently.
I've been seeing this when running the integration tests locally, again the `master` branch:

``` bash
A8_TEST_K8S=false make test.integration
```

If the test fails, check the logs of the bookinfo/gateway containers.
You can also wait a few more minutes, so that more updates will come from the registry (due to productpage cycling through register/unregister).

**Comment by zcahana** [comment]
Resolved by #421
